src/metabase/models/search_index_metadata.clj

```diff
@@ -0,0 +1,81 @@
+                         :engine engine
+                         :version version
+                         :status "pending")
+     (try
+       (t2/insert! :model/SearchIndexMetadata {:engine     engine
+                                               :version    version
+                                               :status     "pending"
+                                               :index_name (name index-name)})
+       true
+       (catch Exception _
+         ;; We assume that failure corresponds to a unique index conflict (a pending entry already exists)
+         false)))))
+
+(defn delete-index!
+  "Delete the given pending index, as long as its still pending."
+  [engine version index-name]
+  (t2/delete! :model/SearchIndexMetadata :engine engine :version version :index_name (name index-name)))
+
+(defn active-pending!
+  "If there is 'pending' index, make it 'active'. Return the name of the active index, regardless."
+  [engine version]
+  (t2/with-transaction [_conn]
+    (when (t2/exists? :model/SearchIndexMetadata :engine engine :version version :status "pending")
+      (t2/delete! :model/SearchIndexMetadata :engine engine :version version :status "retired")
+      (t2/update! :model/SearchIndexMetadata {:engine engine :version version :status "active"} {:status "retired"})
+      (t2/update! :model/SearchIndexMetadata {:engine engine :version version :status "pending"} {:status "active"}))
+    (t2/select-one-fn :index_name :model/SearchIndexMetadata :engine engine :version version :status "active")))
+
+(defn delete-obsolete!
+  "Remove metadata corresponding to obsolete Metabase versions.
+  It is up to the relevant engine to delete the actual indexes themselves."
+  [our-version]
+  ;; If there are no recent versions, then there is nothing to delete.
+  (when-let [most-recent (seq (map :version (t2/query {:select   [:version]
+                                                       :from     [(t2/table-name :model/SearchIndexMetadata)]
+                                                       :group-by [:version]
+                                                       ;; use pk as a tie-breaker
+                                                       :order-by [[[:max :updated_at] :desc]
+                                                                  [[:max :id] :desc]]
+                                                       :limit    3})))]
+    (t2/query-one {:delete-from [(t2/table-name :model/SearchIndexMetadata)]
+                   :where       [:or
+                                 [:not-in :version most-recent]
+                                 ;; Drop those older than 1 day, unless we are using them, or they are the most recent.
+                                 [:and
+                                  [:not-in :version (filter some? [our-version (first most-recent)])]
+                                  [:<= :updated_at
+                                   (case (mdb/db-type)
```

Comment:
Not sure what you mean 😅

Reply:
`[:<= :updated_at (t/minus (t/now) 1 :hour))]` or whatever the syntax is :)

Reply:
Oh, yeah I guess we don't need precise db time and so that's much simpler. It would also have made testing easier too. Doesn't feel worth another CI run though 😊

Reply:
> Doesn't feel worth another CI run though

this sucks so much.